File: src/network/lobby/packets/user/ItemUserPacket.java
package network.lobby.packets.user;

import misc.Debug;
import network.lobby.packets.UserPacket;
import network.lobby.packets.PlayerPropertyUserPacket;
import player.property.item.Item;
import player.property.PlayerProperty;

public class ItemUserPacket extends UserPacket implements PlayerPropertyUserPacket
{
	private byte[] itemIDs;

	public ItemUserPacket(byte[] itemIDs)
	{
		this.itemIDs = new byte[3];
		for (byte i = 0; i < itemIDs.length; i++)
		{
			this.itemIDs[i] = itemIDs[i];
		}
	}

	public ItemUserPacket(ItemUserPacket packet)
	{
		this(packet.getItemIDs());
	}

	public void assign(ItemUserPacket itemPacket)
	{
		Debug.warnIf(itemPacket == null, "ItemUserPacket.assign(null)");
		itemIDs = itemPacket.itemIDs;
	}

	// Setter
	@Override public void setIDs(byte[] ids)
	{
		for (int i = 0; i < ids.length; i++)
		{
			itemIDs[i] = ids[i];
		}
	}

	// getter
	public Item[] getItems()
	{
		Item[] items = new Item[3];
		for (byte i = 0; i < 3; i++)
		{
			items[i] = Item.getByID(itemIDs[i]);
		}
		return items;
	}

	public byte[] getItemIDs() { return itemIDs; }

	@Override public PlayerProperty[] getPlayerProperty() { return getItems(); }
}

File: cpp/src/network/packets/ItemUserPacket.cpp
#include "ItemUserPacket.hpp"

#include "../../player/property/item/Item.hpp"
#include "../../misc/Debug.hpp"

ItemUserPacket::ItemUserPacket(const std::string& ids)
{
	setIDs(ids);
}

ItemUserPacket::ItemUserPacket(CompressBuffer* buffer)
	: itemIDs(buffer->cut(3))
{}

ItemUserPacket::ItemUserPacket()
	: itemIDs({-1, -1, -1})
{}

bool ItemUserPacket::isValid() const
{
	return itemIDs[0] != -1 && itemIDs[1] != -1 && itemIDs[2] != 1;
}

const std::vector<PlayerProperty*> ItemUserPacket::getPlayerProperties() const
{
	std::vector<PlayerProperty*> props;
	for (int i = 0; i < 3; i++)
	{
		if (itemIDs[i] == -1)
		{
			props.push_back(NULL);
		}
		else
		{
			props.push_back(Item::get(itemIDs[i]));
		}
	}
	return props;
}

void ItemUserPacket::setIDs(const std::string& ids)
{
	itemIDs = ids;
}

std::string ItemUserPacket::toString() const
{
	std::string s;
	for (int i = 0; i < 3; i++)
	{
		s += itemIDs[i];
	}
	return s;
}

CID ItemUserPacket::getCID() const
{
	return ITEM_USER_PACKET_CID;
}
